### ISSUE TITLE
Default values + destructuring confuses CFA 

### ISSUE THREAD

**Issue by jwbay** [Microsoft/TypeScript#12620]
**TypeScript Version:**  2.1.1 and nightly (2.2.0-dev.20161202)

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
let node: {
    props: any;
};

const { props: { className = '' } = {} } = node;

if (className) {
    node.props.className = className.slice(0, 1); // error TS2339: Property 'slice' does not exist on type 'never'.
}
```

**Expected behavior:**
Code compiles without error as in TS 2.0.10

**Actual behavior:**
Error as commented.

Might be duplicate/related to #10065 and its linked issues somehow, but I'm not sure. Strict nullchecks is not enabled.

The following workarounds fix the error:
1. Change const to let
2. Use `if typeof className === 'string'` instead of `if (className)`
3. Take out the default values from the destructuring pattern

**Comment by mhegazy** [comment]
`className` should be `any` really, and i think this is the issue. we give `className` the type from the intializer, and that is not correct, it should be `typeof initializer | typeof node.props.className`

**Comment by mhegazy** [comment]
here is an example to demostrate the issue:

```ts
let node: {
    props: any;
} = { props: { className: null } };

const { props: { className = '' } = {} } = node;

if (className)  // this should be a valid check, since className can be null.
```

**Comment by manojdcoder** [comment]
Looks like issue I'm facing is similar,

```
// default value for b can't be a string
function call({a, b = "Xyz"}: { a: string, b?: number }) {
    alert(b);
}

call({
    a:'Hello'
})

call({
    a: 'Hello',
    b: 10
})
```

**Comment by sandersn** [comment]
Fix is up at #18955.
